### DISCUSSION TITLE
Feature Requests (minor)

### DISCUSSION THREAD

**Discussion by tefkah** [org-roam/org-roam-ui#8] [Feature request]
Please post your *minor* feature requests down below. With minor we mean features that wouldn't change the way org-roam-ui works drastically, but make the experience all around more pleasant. Think of being able to set your own colors or being able to filter by tags.

Also, please check whether we are not already working it [on the project board](https://github.com/org-roam/org-roam-ui/projects)!

**Comment by shaunsingh** [comment]
Would it be possible to add some basic xwidgets support? I.e. add a command to open org-roam-ui in an xwidgets window, and launch org-mode-ui if it isn't open already? Should be as simple as telling xwidgets to open the link

**Reply by tefkah** [reply]
It should be possible, however I'm not running an Emacs with xwidgets support (I think), so I would not be able to test it myself! If someone would be able to add it I'd happily merge it.
I have tried using it with EAF, but since I'm running the PGTK build it's rather buggy, but otherwise that would be a solution.

**Reply by shaunsingh** [reply]
![image](https://user-images.githubusercontent.com/71196912/127782304-f6b3ed92-fe9e-49a5-9ee2-2dbf1e03b116.png)

Opening the link manually works just fine with xwidgets. I don't have much elisp knowledge, but I'll see if I can write something to have xwidgets open the link with a function. After that it should be as simple as adding it to [org-roam-ui.el](https://github.com/org-roam/org-roam-ui/blob/main/org-roam-ui.el) right?

**Reply by tefkah** [reply]
Yes should just work! And dang, that does look cool, how's the performance using xwidgets?

**Reply by shaunsingh** [reply]
The performance is pretty good, compared to chrome it feels just as responsive. Of course, this is only using two nodes and I'm on a fairly powerful machine. 

However, I did run this: https://vasturiano.github.io/react-force-graph/example/large-graph/, and it was just as fast as chrome and safari. 

I'm also mostly done with the minor mode that 
1. opens org-roam-ui in a popup buffer if you do have xwidgets or 
2. opens org-roam-ui in your default browser if you don't

Should be much easier than opening that long url every time :p

**Reply by shaunsingh** [reply]
Apparently the latest branch opens the url by default, and if you have xwidgets as your browser, it works just as well! 

By any chance, is it possible to set default settings in config.el/init.el? For example if I want 3d on by default, can I set it in my config?

**Reply by tefkah** [reply]
Great to hear that that works!

No, not yet. We have been thinking about providing a complete interface into the settings from emacs, which would not be tooo difficult to make, but unless a lot of people want that I won't implement it yet as that will most likely break something somewhere haha.

What I could do is make the 3D position be remembered in the same way as the other settings, so you don't have to set it anyway!

**Reply by tefkah** [reply]
Latest build should have 3d be remembered across sessions

**Comment by tecosaur** [comment]
###### Disclaimer: I'm not sure if this is where you want this sort of thing, please let me know how you want to organise suggestions like this if it's not here :slightly_smiling_face: 

I've recently started using this, and you've all done a great job! It looks fantastic. I have a few minor niggles though which I'll put below.

## I think a text shadow would be nicer than a background

I see that when I hover over a node the text is given a background to make it stand out better

![image](https://user-images.githubusercontent.com/20903656/127778583-00151418-1146-4efc-8cc3-eceaebd286fa.png)

I think this works well with the big "Keyboard Layouts" tooltip, but with the annotations on the nodes: having worked on something actually rather similar to this before I found that I was able to achieve similar text clarity but with a cleaner overall look overall. Is this something that you've considered?

## Not enough indication of vertical overflow with the menu

This is something that tripped me up for a moment.

<img src="https://user-images.githubusercontent.com/20903656/127778684-75e7c96d-97ae-4b58-a5b5-3019d8a8380a.png" height="400em">

Some expansions of the menu fit in the vertical space really well (see above) and so in the absence of clear visual indicators, it doesn't instinctually *feel* like I should scroll down to see more.

In my mind, something like a little chevron would go a long way (shoddily drawn on with my screenshot tool)

<img src="https://user-images.githubusercontent.com/20903656/127778863-b041b9f6-5108-4622-8747-7123fd1ceb9d.png" height="150em">

Lastly, when there's a lot of vertical space, I think it would be nice if the menu expanded a bit before the scroll kicked in, if possible.

<img src="https://user-images.githubusercontent.com/20903656/127778912-7e2cb45a-039a-4f06-966e-43b2cde2714d.png" height="500em">

**Reply by tefkah** [reply]
Thanks for the detailed feedback, and I agree on all fronts! I haven't been too happy with the background either (although you can customize them to your liking somewhat) but text shadow seems like such a better solution that I feel silly not having thought of it. Actually, what might work best would be some small blur behind the text combined with some shadow, although I'm not sure how to implement that in HTML canvas (someone probably does thought).

As for the showing/expanding, I agree: the scrolling is somewhat unclear, although I'm not completely sold on the chevron icon (to me it indicates: click here to expand, rather than scroll down) but that might just be me. The box not getting bigger is indeed my fault and mostly due to me giving up trying to set a max height using percentages  and flexboxes :p. I'll try to fix that.

And this is exactly where I would want this sort of thing, keep them coming! Although it might be better to have a "UX suggestions" discussion, although then again, that might incentivize people to give all kinds of U*I* suggestions, where more opinions rarely make a better product haha. But UX wise please keep it coming, it's easy to overlook this kind of thing and I'd like to make ORUI as accessible and intuitive as possible.

**Reply by tefkah** [reply]
https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/shadowBlur

Canvas has, like, 20 native methods, but blurring shadows is exactly one of them, how great!

**Reply by tecosaur** [reply]
> what might work best would be some small blur behind the text combined with some shadow

This is exactly what I ended up doing :) except in my case by using a gradient as a CSS shadow.

> I agree: the scrolling is somewhat unclear, although I'm not completely sold on the chevron icon

I'm not too fussed about the chevron, so much as there being _something_ to indicate more being below the fold.

> set a max height using percentages and flexboxes :p. I'll try to fix that.

Cool!

> UX wise please keep it coming

I'll be sure to plop any other thoughts I have here then :)

**Reply by tefkah** [reply]
I fixed a couple of these things by moving the label down and making the tweaks tool longer. I have tried out textShadow, however, due to the way the library we're using is made, this also adds shadow/blur to the nodes, which sucks. It's also really slow, so I will put it on the backburner for now. Maybe in the future we will fully do custom nodes, but for now we'll stick with this I think. I couldnit figure out a way to visualize "more down here" that I liked yet.

**Reply by tecosaur** [reply]
I find the change nicer, thanks for what you've done :slightly_smiling_face:. One other niggle which is more apparent with the new under-node labels is undesirable duplication:

![image](https://user-images.githubusercontent.com/20903656/128679481-3a351641-ef3d-4c95-af25-00757a56fbef.png)

I think if a node is being hovered over and a tool tip with the node name is present, the normal node label should be hidden. Otherwise, it can look a bit ugly.

![image](https://user-images.githubusercontent.com/20903656/128679679-b0b655f1-22dd-4872-a9f3-bdc3d772718b.png)

It may also be worth making sure the tooltip is no smaller than the node labels. This looks a bit odd:

![image](https://user-images.githubusercontent.com/20903656/128679760-075b59c1-d03b-4f4c-bd32-4cb93566b8d0.png)

**Reply by tefkah** [reply]
I also don't really like the Tooltip mixup, but do you think the original label should be hidden or the tooltip? I like Obsidian's behavior here, basically: there is no tooltip, just a label.

https://user-images.githubusercontent.com/21983833/128700216-06a927e7-40f0-492d-ada6-8058151793f6.mp4



On that note (haha, notes): what do you think about the highlighting behavior? Currently it shows all the titles of the surrounding notes as well, which can end up being kind of cluttered. On the other hand, I don't really like Obsidians approach where you only see the title of the current node, as that gives you very little idea to what it's linked to. Maybe making the other highlighted labels smaller would work?

On more thing: Currently the labels are all on one line as that is really easy to do. However, sometimes I would like to see the entire title but not have it clutter the whole screen, so maybe word-wrapping the title would be an option. I've looked into how to do this in HTML canvas and it should be possible, but it will take some time and I wanted to check whether people care about this before I try to implement it :p

**Reply by tecosaur** [reply]
I'm thinking just show the label, but make the label for the hovered over / active node more prominent somehow (minimum font size, 1.2x the normal label font size, slightly more contrast against the background, a tad bolder, ...).

Regarding word wrapping, I'm not really sure whether it would be worth the effort - I can't say it's a problem I've run into yet.

**Comment by tecosaur** [comment]
### Re: Settings in emacs
I thought it would be best to start a new thread on this

> We have been thinking about providing a complete interface into the settings from emacs, which would not be tooo difficult to make, but unless a lot of people want that I won't implement it yet as that will most likely break something somewhere haha.

What sort of data structure is the current JS settings object?

**Reply by blester125** [reply]
I personally would like the ability to tweak all the setting in emacs, when the setting are done in my emacs config they can be checked into version control and shared across computers. The current method of adjusting it in the UI is nice, but, given that it is saved into local storage, it requires re-setting them on different computers, browsers, etc.

I think setting defaults in emacs, or at least some way to serialize and load the settings from disk, would really help create a consistent experience for anyone who uses it across different computers.

**Reply by tecosaur** [reply]
Ah, I see [components/config.ts](https://github.com/org-roam/org-roam-ui/blob/main/components/config.ts). This looks like it would be pretty easy with `json-encode-plist` merged (on the JS side) with the default properties using the spread operator.

**Reply by tefkah** [reply]
> I personally would like the ability to tweak all the setting in emacs, when the setting are done in my emacs config they can be checked into version control and shared across computers. The current method of adjusting it in the UI is nice, but, given that it is saved into local storage, it requires re-setting them on different computers, browsers, etc.
> 
> I think setting defaults in emacs, or at least some way to serialize and load the settings from disk, would really help create a consistent experience for anyone who uses it across different computers.

This is a good point which I had failed to take into account, I kind of assumed it would be a "set-it-and-forget-it" kind of thing for most people, but of course some people actually go outside and might need to use different computers :p  

Implementing it should not be very difficult, we will add it eventually!

The settings are just encoded in a few very simple json objects, so it should not be too hard to transfer them.

**Comment by blester125** [comment]
Now that cite links appear in the graph (I love the the new customizations for cites—setting the cite node color, cite link color, and the dashed citelinks—by the way), it would be nice if those links (and the cite nodes) were highlighted when you mouse over node they are connected to, preferably with the ability to customize these highlight colors separately from the defaults (although I realize all these separate customization options for cite links are probably causing a lot of copy-paste-modify code).

**Reply by tefkah** [reply]
Hmmmm that could be done, would you want a separate highlighting color for them or for them to retain their original color when highighted? They dont do the latter yet, but that should be doable, the former will be a bit of a pain in the ass, but the way we do those checks needs to be modified at some point anyway, so well probably get to it

**Reply by tefkah** [reply]
I said it was a pain, and it is a little, but I added it anyway! I will add it for your links as well.

**Comment by thriveth** [comment]
Allow choice of browser. 
I use Firefox for my default browser, but would like to be able to run ORUI in Chromium. It would be great to be able to define my own "open browser" command.

**Reply by tefkah** [reply]
Added! If you do 
```emacs-lisp
(setq org-roam-browser-function #'browse-url-chromium) 
```
you should be all set

**Reply by thriveth** [reply]
Fast and efficient as always! Thank you so much!
